vmware_content_deploy_template: Make parameter optional

Akasurde opened the pull request:
##### SUMMARY

* Marked Datastore, Hostname, Resource pool name, and cluster name as optional
* Added appropriate check, doc changes, and a warning message
* User should either has to specify cluster name or hostname with the resource pool name
* Added Datastore Cluster parameter

Fixes: #397 #550

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>


##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
changelogs/fragments/397_vmware_content.yml
docs/community.vmware.vmware_content_deploy_template_module.rst
plugins/modules/vmware_content_deploy_template.py
plugins/module_utils/vmware.py
plugins/modules/vmware_guest.py
plugins/modules/vmware_guest_cross_vc_clone.py

Akasurde commented:
@ultral Can you review this? This is similar to other PR.

ultral commented:
> @ultral Can you review this? This is similar to other PR.

sure, I'll review it

Akasurde commented:
@MalfuncEddie Could you please check the update?

MalfuncEddie commented:
> @MalfuncEddie Could you please check the update?



> @MalfuncEddie Could you please check the update?

i'll test it

MalfuncEddie commented:
Hi,

Template is deployed with only cluster name (y)


But the datastore cluster parameter still gives an error:
`line 204, in deploy_vm_from_template\nAttributeError: 'VmwareContentDeployTemplate' object has no attribute 'datastore_id`

Also reading this PR

User either has to specify cluster name or hostname with the resource pool name 
--> should be  cluster name or hostname. 
A resource pool name should not be required. If not defined it should use the 'user invisible' resourcepool "resources".

Akasurde commented:
> Hi,
> 
> Template is deployed with only cluster name (y)
> 

good


> But the datastore cluster parameter still gives an error:
> `line 204, in deploy_vm_from_template\nAttributeError: 'VmwareContentDeployTemplate' object has no attribute 'datastore_id`
> 

modified PR.

> Also reading this PR
> 
> User either has to specify cluster name or hostname with the resource pool name
> --> should be cluster name or hostname.
> A resource pool name should not be required. If not defined it should use the 'user invisible' resourcepool "resources".

modified description

mariolenz commented:
recheck

mariolenz commented on changelogs/fragments/397_vmware_content.yml:
```diff
@@ -0,0 +1,3 @@
+breaking_changes:
```
```suggestion
minor_changes:
```

I don't think this is a breaking change since it shouldn't really _break_ existing playbooks. Going from optional to required could, but not the other way round imho. Or did I overlook something?

@goneri What do you think? I'd like to avoid a situation like #540

goneri replied:
Thank you @mariolenz for raise this. Yes, I would also prefer we avoid any breaking change. Or this means we need to open up a 2.x branch.

mariolenz replied:
@Akasurde 

Is there a reason why you think this is a breaking change? As I've said above, I don't think so. (Going from optional to required shouldn't break anything.)

If it really _does_ break existing playbooks we can't ship this in 1.6, we would need a major version 2.0. And _this_ would mean, as far as I know, that it will never end up in ansible 2.10 because this version will only ship community.vmware 1.x.

cc @goneri

MalfuncEddie commented:
Hi

Two remarks, should vmware_content_deploy_ovf_template not follow the same logic as defined in this PR for vmware_content_deploy_template  ? 
And what (except of the template type) is the difference between vmware_content_deploy_ovf_template and vmware_content_deploy_template ?

Shouldn't this be the same module with an extra parameter template type?

mariolenz commented:
> Should vmware_content_deploy_ovf_template not follow the same logic as defined in this PR for vmware_content_deploy_template ? (cluster or hostname, datarore of datastore cluster ...)

Yes, definitely.

> And what (except of the template type) is the difference between vmware_content_deploy_ovf_template and vmware_content_deploy_template ? Shouldn't this be the same module with an extra parameter template type?

Good question. There may be good reasons to have two modules. @pgbidkar @ultral What do you think? You're the original authors of both modules, so maybe you can help us here. Do you think they can be merged, or are there good reasons to keep them separated?

> Datastore Clusters would also be a great addition to vmware_content_deploy_ovf_template

I'm not 100% sure but iirc the vCenter API wants to have a datastore, not a datastore cluster. But I'd have to have a closer look at this. If the API allows to define a datastore cluster, I agree that the module should support this also. Btw: Beeing able to define a storage policy would also be nice ;-)

MalfuncEddie commented:
> > Datastore Clusters would also be a great addition to vmware_content_deploy_ovf_template
> 
> I'm not 100% sure but iirc the vCenter API wants to have a datastore, not a datastore cluster. But I'd have to have a closer look at this. If the API allows to define a datastore cluster, I agree that the module should support this also. Btw: Beeing able to define a storage policy would also be nice ;-)

You are correct that the api expects a datastore. But when you supply a datastore cluster you should query the api for StoragePod and get a recommendation and then use that datastore.  

- determine if datastore is cluster or normal datastore 
--> normal --> deploy
--> datastore cluster --> ask a recommendation --> get datastore --> deploy

mariolenz commented:
@MalfuncEddie 
> You are correct that the api expects a datastore. But when you supply a datastore cluster you should query the api for StoragePod and get a recommendation and then use that datastore.

Good point, didn't think of this. I've found this in `vmware_guest`:

https://github.com/ansible-collections/community.vmware/blob/136052a5c52634ac99b26fc009278e93f3071564/plugins/modules/vmware_guest.py#L2720

This seems to do the trick. Maybe we should move this function to `vmware.py` so other modules like `vmware_content_deploy_ovf_template` can make use of it.

> Not sure I should log an extra issue for this so let me know :)

Yes, I think this discussion is a bit out of scope for this PR. A dedicated issue would be a good idea.

Btw: I've opened #550 in preparation.

MalfuncEddie commented:
Hi,

Could you also make the datastore changes to vmware_content_deploy_ovf_template.py since it could also use the datastore_cluster parameter.

Also in vmware_guest.py they use datastore for datastores and clusters. What would make the code more portable between vcenters that have no datastore clusters.


eg:
If I have 2 vcenters
1 has datastore clusters and the other not

the first vcenter I would need to use datastore_cluster
If I use the same code on vcenter 2 I would have to change datastore_cluster to datastore.

You can use a work around like:
gather facts about datastore clusters
see if datastore cluster exists --> if so set a custom fact set_datacluster
if not set a fact set_datastore

datastore_cluster: "{{ set_datacluster | default(omit) }}"
datastore: "{{ set_datastore | default(omit) }}"


bit of a work around so it would be a lot easier to hqve to ability to just use datastore and let the module figire it out if it is a datastorecluster(and get the recommended) of normal datastore.


Am I making any sense?

mariolenz reviewed (approved):
LGTM

mariolenz commented:
It looks like this PR will also fix #550 :-)

Btw: Is it possible to define more than just one issue in `Fixes:`?
